lua/apisix/plugins/basic-auth.lua

```diff
@@ -0,0 +1,224 @@
+    end
+
+    return true
+end
+
+local function extract_auth_header(authorization)
+
+    local function do_extract(auth)
+        local obj = { username = "", password = "" }
+
+        local m, err = ngx.re.match(auth, "Basic\\s(.+)")
+        if err then
+            -- error authorization
+            return nil, err
+        end
+
+        local decoded = ngx.decode_base64(m[1])
+
+        local res
+        res, err = ngx_re.split(decoded, ":")
+        if err then
+            return nil, "split authorization err:" .. err
+        end
+
+        obj.username = ngx.re.gsub(res[1], "\\s+", "")
+        obj.password = ngx.re.gsub(res[2], "\\s+", "")
+        core.log.info("plugin access phase, authorization: ", obj.username, ": ", obj.password)
+
+        return obj, nil
+    end
+
+    local matcher, err = lrucache(authorization, nil, do_extract, authorization)
+
+    if matcher then
+        return matcher.username, matcher.password, err
+    else
+        return "", "", err
+    end
+
+
+end
+
+function _M.access(conf, ctx)
+    core.log.info("plugin access phase, conf: ", core.json.delay_encode(conf))
+
+    -- 0. check the plugin is enabled
+    if not conf.enable then
+        return
+    end
+
+
+    -- 1. extract authorization from header
+    local headers = ngx.req.get_headers()
+    if not headers.Authorization then
+        return 401, { message = "authorization is required" }
+    end
+
+    local username, password, err = extract_auth_header(headers.Authorization)
+    if err then
+        return 401, { message = err }
+    end
+
+    -- 2. get user info from etcd
+    local res = authorizations_etcd:get(username)
+    if res == nil then
+        return 401, { message = "failed to find authorization from etcd" }
+    end
+
+    -- 3. check user exists
+    if not res.value or not res.value.id then
+        return 401, { message = "user is not found" }
+    end
+
+    local value = res.value
+
+    -- 4. check the password is correct
+    if value.password ~= password then
+        return 401, { message = "password is error" }
+    end
+
+
+    core.log.info("hit basic-auth access")
+end
+
+
+local function set_auth()
+    local body_table = {}
+    -- read_body can not use in log_by_lua
+    if ngx.re.find(ngx.req.get_headers()["Content-Type"] or "", "application/json") then
+        ngx.req.read_body()
+
+        local body_data = ngx.req.get_body_data()
+        if body_data ~= nil then
+            body_table = json.decode(body_data)
+        end
+
+    else
+        body_table = ngx.req.get_post_args()
+    end
+
+    local username = body_table["username"]
+    local password = body_table["password"]
+
+    if not username or not password then
+        core.response.exit(200, "username,password is required")
+    end
+
+    local key = gen_key(username)
+
+    local res, err = core.etcd.set(key, { username = username, password = password })
+    if not res then
+        core.response.exit(500, err)
+    end
+
```

Comment:
plugin don't know the exist of etcd, which should be deal with admin API and core.